### ISSUE TITLE
Span<T> needs to be support long length and indices

### ISSUE THREAD

**Issue by GPSnoopy** [dotnet/corefxlab#896]
Half of the time we use native memory is to break the 32-bits barriers. Designing a new API in 2016 that is not ready for 64-bits is questionable.

**Comment by KrzysztofCwalina** [comment]
@jkotas, @jaredpar, @terrajobst, @davidfowl 

This is a good question/issue to discuss. But I do worry around making the change. All indices in the Framework are typed as Int32, and having one type in the Framework that is different would mean lots of often annoying casts. Also, it might make Span\<T\> larger (more costly to pass by value)

**Comment by jkotas** [comment]
If we were to do this change, we should do it right and make the indices native ints (ie 32-bit on 32-bit platforms, 64-bit on 64-bit platforms). It would require better support for native ints in C# that is another popular topic: https://github.com/dotnet/roslyn/issues/2788, https://github.com/dotnet/coreclr/issues/963, https://github.com/dotnet/corefx/issues/10457, ...

**Comment by omariom** [comment]
I would vote for postponing a bit pattern matching improvements and implementing native ints instead :disappointed: 
/cc @VSadov

**Comment by benaadams** [comment]
The span pointer is 64 bit even if its length is only 32 bit so it allows you to process the memory in 2GB slices. Making it all 64bit will increase its size and the registry pressure and copying cost between functions, for no real gain.

`Span<T>` is a stack only type so unless you a running through > 2GB all at once in a single function; why would it need to be 64 bit? You can chunk the memory into multiple span chunks as you move through it. The additional cost of slicing int to 2GB chunks would be fairly minimal.

`Memory<T>` on the other hand, which is a heap wrapper, maybe a `LargeMemory<T>` type?

**Comment by omariom** [comment]
@benaadams It could be 8 bytes on x86 and 16 on x64. x64 RyuJit copies such structs using single xmm register. Though I agree that 2 GB is enough in 99% of scenarios.

**Comment by KrzysztofCwalina** [comment]
We decided to keep it as an int. The reasoning is described in https://github.com/dotnet/apireviews/tree/master/2016/11-04-SpanOfT

**Comment by vadimkantorov** [comment]
Be it a ROS or something else, there is big demand for arbitrary-length memory wrappers for interfacing with HPC and Deep Learning applications. I wrote up P/Invoke bindings for DLPack tensors (a standard tensor interchange format between PyTorch / TensorFlow / MXNet etc): https://github.com/vadimkantorov/dotnetdlpack/blob/master/dlpack.cs#L85

I had to call the method DataSpanLessThan2Gb because of this issue.

For these applications the size of wrapping struct is not as important. An alternative would be a faithful another type for wrapping memory.
